compiler/rustc_lint/src/context.rs

```diff
@@ -419,29 +419,49 @@ impl LintStore {
         }
     }
 
-    fn no_lint_suggestion(&self, lint_name: &str) -> CheckLintNameResult<'_> {
+    fn no_lint_suggestion(&self, lint_name: &str, tool_name: &str) -> CheckLintNameResult<'_> {
         let name_lower = lint_name.to_lowercase();
 
         if lint_name.chars().any(char::is_uppercase) && self.find_lints(&name_lower).is_ok() {
             // First check if the lint name is (partly) in upper case instead of lower case...
-            return CheckLintNameResult::NoLint(Some(Symbol::intern(&name_lower)));
+            return CheckLintNameResult::NoLint(Some((Symbol::intern(&name_lower), false)));
         }
+
         // ...if not, search for lints with a similar name
-        // Note: find_best_match_for_name depends on the sort order of its input vector.
-        // To ensure deterministic output, sort elements of the lint_groups hash map.
         // Also, never suggest deprecated lint groups.
-        // We will soon sort, so the initial order does not matter.
         #[allow(rustc::potential_query_instability)]
-        let mut groups: Vec<_> = self
+        let groups: Vec<_> = self
             .lint_groups
             .iter()
             .filter_map(|(k, LintGroup { depr, .. })| depr.is_none().then_some(k))
             .collect();
-        groups.sort();
         let groups = groups.iter().map(|k| Symbol::intern(k));
         let lints = self.lints.iter().map(|l| Symbol::intern(&l.name_lower()));
-        let names: Vec<Symbol> = groups.chain(lints).collect();
-        let suggestion = find_best_match_for_name(&names, Symbol::intern(&name_lower), None);
+        let mut names: Vec<Symbol> = groups.chain(lints).collect();
+        // Here, a little bit of extra hack
+        // we use the lint names from rustc and to mock the tool name,
+        // if it's selected we then strip to the right suggestion with `true` for from_rustc
+        // so we can handle these cases:
+        //  1. suggest `missing_docs` from rustc for `clippy::missing_docs`
+        //  2. suggest `dead_code` from rustc for `clippy::dead_cod`
+        let rustc_names = self
+            .by_name
+            .keys()
+            .map(|k| Symbol::intern(&format!("{tool_name}::{k}")))
+            .collect::<Vec<_>>();
+        names.extend(rustc_names.clone());
+        // Note: find_best_match_for_name depends on the sort order of its input vector.
+        // Sort elements here to ensure deterministic output
+        names.sort();
+        let suggestion =
+            find_best_match_for_name(&names, Symbol::intern(&name_lower), None).map(|s| {
```

Comment:
I'd do this a little differently, first call `find_best_match_for_name` with only the tool lints, and then if that fails, strip the tool name from `name_lower` and then call `find_best_match_for_name` again. That seems tidier than changing all of the suggestion names and then stripping the tool name off them again afterwards.

Reply:
https://github.com/rust-lang/rust/pull/119819#discussion_r1448618397

Seems like you were 6 minutes faster than me :sweat_smile:

Reply:
This change is also trying to fix the issue in this comment: https://github.com/rust-lang/rust/pull/119819#discussion_r1447650717.

The problem in your solution is, for `clippy::dead_cod` as lint name, 
 `find_best_match_for_name` will first return `Some` with `clippy::eq_op` from tool lints, since it's the closest one if we call `find_best_match_for_name` without limit.

How do we define it's failed? 

I thought there is no lints names from `rustc` in `names`, I just double checked we have them in it! yeah, we don't need to extend names.

it will be great if `find_best_match_for_name` may return the edit distance with returned string, so we may compare the result of `find_best_match_for_name(names, tool_name, None)` and `find_best_match_for_name(names, tool_name_stripped, None)`, then choose the closer one.

Reply:
or maybe we can first try with a limit such as `find_best_match_for_name(names, tool_name, Some(2))`, if anything is found return the result, otherwise try `find_best_match_for_name(names, tool_name_stripped, None)`.

Reply:
Hmm, that is tricky. You can modify `find_best_match_for_name`'s interface if there's an easy way to give it a maximum edit distance or return the edit distance that the result had (perhaps keep `find_best_match_for_name` the same and add another function with a more complicated interface - something like this might already exist).

Reply:
I agree, but we need to get some balance here, I want to avoid some confusion:

`clippy::missing_docs` suggest to `clippy::erasing_op`, _a better candidate is to `missing_docs`_.
`clippy::dead_cod` suggest to `clippy::eq_op`, _a better candidate is to `dead_code`_.

Users may don't know whether a lint is from Clippy or rustc, and only know part or full of the lint's name.

So we need to point out `from_rustc` when the suggested lint is not from `clippy`.

I add a new API `find_best_match_for_names` for this scenario.